### ISSUE TITLE
EVENTLET_NO_GREENDNS should still be available

### ISSUE THREAD

**Issue by petrutlucian94** [eventlet/eventlet#390]
The 'EVENTLET_NO_GREENDNS' option has been deprecated, dnspython being enforced. Users may want to stick with the native dns resolution method, considering issues that dnspython may have (e.g. https://github.com/rthalley/dnspython/issues/240).

**Comment by temoto** [comment]
Yup, it's back as part of fixing #383 
merged in master 78f2ef8a81600d19d6d670803ef73747ab8d81b9
will be released on PyPI soon.

**Comment by temoto** [comment]
Use following construct if you need it now.

`pip install https://github.com/eventlet/eventlet/archive/78f2ef8a81600d19d6d670803ef73747ab8d81b9.zip`

**Comment by petrutlucian94** [comment]
That's great, thanks for the quick reply.

**Comment by kashirin-alex** [comment]
The 20.1 version includes the dnspython package at /eventlet/support/
Can it be possible to stick to the dependency ? otherwise, it is almost as required to have it in a new package name or delete each time and version control of dnspython probably would be unparalleled. 

version 0.9.10 was:
> Greendns: if dnspython is installed, Eventlet will automatically use it to provide non-blocking DNS queries. Set the environment variable ‘EVENTLET_NO_GREENDNS’ if you don’t want greendns but have dnspython installed.


Thanks,
Kashirin Alex

**Comment by temoto** [comment]
@kashirin-alex I don't understand about version control. Do you experience problems?

**Comment by kashirin-alex** [comment]
The main cause I expected the dnspython (on an issue to be from the dnspython package) while It suddenly it was in the eventlet path, in my case, I tried to allocated the 'method _validate_labels'
and it is different package versions, one at line 241 other's at 273
https://github.com/eventlet/eventlet/blob/master/eventlet/support/dns/name.py#L241
https://github.com/rthalley/dnspython/blob/master/dns/name.py#L273
so, to keep it aligned it is probably have it as dependency. I have important usage of the dnspython as it is being the lib for the name-servers.

second matter about the dnspython wrapper for the resolver, there are issues of resolutions when the name is '' empty(ipv4 only) , '::' ipv4+ipv6 , 'localhost' , It tries any way to resolve to an ip and on usage of socket.bind

Both things are not urgent for me, while, everything was good, and suddenly these issues arise with the 20.1 release and I thought It would be great to check on.

Thank You,
Kashirin Alex

**Comment by temoto** [comment]
@kashirin-alex it is still hard to understand the problem. Can you describe it in terms of answering these questions? Preferably in a separate issue, because what you say sounds like not related to environment variable here.

* I want to have ...
* so I have done ...
* and expected ...
* but instead got ...

**Comment by kashirin-alex** [comment]
The relating to the environment variable, It was good as with 0.9.10, and the dnspython package was not  included

version 0.9.10 was:
> Greendns: if dnspython is installed, Eventlet will automatically use it to provide non-blocking DNS queries. Set the environment variable ‘EVENTLET_NO_GREENDNS’ if you don’t want greendns but have dnspython installed.

the others are different subject issues, in short list:
- hostnames with value of ''empty or '::' not to be resolved
- hostnames not ending localhost, not to be added with the host's zone

**Comment by temoto** [comment]
Thank you, it's clear now.

I agree it's stupid to resolve empty names.

On localhost, please explain why.

**Comment by kashirin-alex** [comment]
regarding
- hostnames not ending localhost, not to be added with the host's zone
a bit further rephrasing, 
- hostnames which are not ending with host's name parent zone not to be resolved with the host's parent-zone (as being the sub-domain short form of hostname) 
or a simpler phrasing of conditions:
 1.  the  /etc/hosts apply only to parent-zones that are equal 
 2.  or exceptionally to localhost, 
 3.  otherwise try to resolve with the dnspython  on the server's default nameservers 


at current state, a request to resolve google.com, the question request is being "google.com.thither.direct"
while if it would be mx.thither.direct it should have been asking for mx.thither.direct and not "mx.thither.direct.thither.direct" as being now ( domain-to-resolve.tld.thither.direct ).
the same apply with the localhost, It tries to resolve localhost.thither.direct

basically, in summary, there are 2 issues on this:
1) hostname to resolve should not be changed in ant way
2) name-servers to use for resolving cannot be the host's parent-zone and these should be the server's default name-servers

Thank You,
Kashirin Alex

**Comment by temoto** [comment]
I understand what's going on with unqualified names resolving. AFAIK, it works as it should. You still has not explained why it should change,

Why it works, IMHO, properly: when you ask to resolve unqualified name, such as `somehost` **and** have a `domain thither.direct` in `/etc/resolv.conf`, then resolver must send `somehost.thither.direct` query to nameservers and optionally also `somehost` query. If you are sure that your query is full domain, write it in FQDN form `somehost.thither.direct.` (trailing dot). In that case resolver will not ask `somehost` without configured domain suffix.

AFAIK, there are no special case exceptions to name `localhost`. Please point to RFC or similar document where it states so.

**Comment by kashirin-alex** [comment]
I did not know about the case of dot
as suffix being the final destination on the resolve request , great to
know.

As for my resolv.conf it is only the NS IPS while the /etc/hosts is in all
fqdn, local loops and the localhost.
I'll keep a check on and as well as for localhost.

Most import is the empty and the ipv6 empty use case that you agree on and
regarding the EVENTLET_NO_GREENDNS, just could be OK to keep of in case
when specificlly to work without greendns (as for the dnspython pkg
included , it is an issue on versioning )

Thanks ,
Kashirin Alex

**Comment by temoto** [comment]
I still don't get the issue on versioning.

**Comment by kashirin-alex** [comment]
if there is already an installed dnspython and it is used else where than
eventlet it might get the pkg in the eventlet instead of the original pkg
install

**Comment by temoto** [comment]
It's not `might`. Eventlet precisely uses only bundled version because what's installed may be incompatible.

**Comment by kashirin-alex** [comment]
in this case the bundled version should be used as eventlet.support.dns and
the original version will be always up to date.(unless other method of
import done only within the support/ path.

The might refers more to the other dnspython libs usages than to the
eventlet's and because the import of the pkg done by path add it overwrites
the original pkg which should remain available for usage.

**Comment by temoto** [comment]
I agree, `import eventlet.support.dns`, is much better. Send patch to dnspython to convert their imports to relative form and I'd be very happy to change our side.

`sys.path` is restored to original after importing bundled dns, so original package is quite available. Read #341 for details.

You are using a lot of imperatives without showcasing the problem they ought to fix. This is not way to go.

Please, describe concrete problems in separate issues. This one is about restoring ability to disable bundled dnspython via environment variable. All our talk above is not relevant to this issue.

**Comment by dianaclarke** [comment]
A number of different OpenStack projects were setting `EVENTLET_NO_GREENDNS`, and are seeing failing tests & behaviour changes since upgrading from `eventlet===0.19.0` to `eventlet===0.20.1`:

- http://codesearch.openstack.org/?q=EVENTLET_NO_GREENDNS&i=nope&files=&repos=

@temoto When you have a chance, can you please cut another release for the change in 78f2ef8a81600d19d6d670803ef73747ab8d81b9? 

- dns: EVENTLET_NO_GREENDNS option is back, green is still default (78f2ef8a81600d19d6d670803ef73747ab8d81b9)

Thanks a bunch!

**Comment by temoto** [comment]
@dianaclarke no wuckers. Please test today's master (de06878) to speed up release.

**Comment by emonty** [comment]
@temoto I've got a patch up now testing that sha - thanks!

**Comment by emonty** [comment]
@temoto @dianaclarke we did some further testing and it seems that releasing eventlet to pick up this fix actually won't fix the things we thought it would - we'll need to dig in on . Thanks for being responsive!

**Comment by dianaclarke** [comment]
@emonty @temoto Indeed, apologies for not being more verbose earlier. This patch (78f2ef8a81600d19d6d670803ef73747ab8d81b9) fixes 15 of the 17 test failures I'm seeing in Designate.

- OpenStack Designate (DNSaaS)
- https://github.com/openstack/designate

And if I recall correctly, a full revert to `eventlet===0.19.0` fixes all the failing tests. If I have some time later, I'll try to dig deeper into the remaining failing tests and report back.

Thanks again for being so responsive. Cheers!

**Comment by temoto** [comment]
No new errors though?

**Comment by emonty** [comment]
Correct - no new errors - the sha you posted works great.

**Comment by dianaclarke** [comment]
I just did another round of test runs locally (just to be sure).

- py27, eventlet==0.20.1, Failed: 0
- py35, eventlet==0.20.1, Failed: 16
- py27, eventlet==0.19.0, Failed: 0
- py35, eventlet==0.19.0, Failed: 0
- py27, eventlet.git@de06878e5a295bfbbddca0048c3453d16168a676, Failed: 0
- py35, eventlet.git@de06878e5a295bfbbddca0048c3453d16168a676, Failed: 0

Interestingly, I'm no longer seeing the 2 additional failures with  eventlet.git@de06878e5a295bfbbddca0048c3453d16168a676 that I was seeing yesterday; my tests all pass now. That's locally though, so perhaps it has something to do with my dev env.

**Comment by temoto** [comment]
v0.21 is uploaded to PyPI. Have an awesome life. 👍

**Comment by emonty** [comment]
Thanks Sergey!!!

**Comment by kashirin-alex** [comment]
Thanks!
